src/lib/core/clock_lowres.c

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright 2010-2022, Tarantool AUTHORS: please see AUTHORS file.
+ *
+ * Redistribution and use in source and binary forms, with or
+ * without modification, are permitted provided that the following
+ * conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above
+ *    copyright notice, this list of conditions and the
+ *    following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above
+ *    copyright notice, this list of conditions and the following
+ *    disclaimer in the documentation and/or other materials
+ *    provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY <COPYRIGHT HOLDER> ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+ * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+ * <COPYRIGHT HOLDER> OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+ * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+ * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+#include <signal.h>
+#include <string.h>
+#include <say.h>
+#include <math.h>
+#include "clock.h"
+
+/**
+ * Granularity in seconds.
+ * SIGALRM handler interrupts some syscalls (open, read, etc.), so
+ * too small granularity is not recommended.
+ */
+#define LOW_RES_GRANULARITY 0.05
+
+double low_res_clock = 0.0;
+
+/** A tick of low_res_clock, SIGALRM handler. */
+static void
+clock_lowres_tick(int signum)
+{
+	(void)signum;
+	low_res_clock += LOW_RES_GRANULARITY;
+}
+
+void
+clock_lowres_signal_init(void)
+{
+	struct sigaction sa;
+	memset(&sa, 0, sizeof(sa));
+	sa.sa_handler = clock_lowres_tick;
+	sa.sa_flags = SA_RESTART;
+	if (sigaction(SIGALRM, &sa, NULL) == -1)
```

Comment:
I believe you forgot to patch the the thread's signal mask.

Reply:
https://github.com/tarantool/tarantool/blob/ac365ecf4607d4345af3e853451fe91d5ee2ff5c/src/lib/core/fiber.c#L1633-L1644

https://github.com/tarantool/tarantool/blob/ac365ecf4607d4345af3e853451fe91d5ee2ff5c/src/tt_pthread.h#L267-L277

Reply:
Let's add `assert(cord_is_main());` to the signal handler then.

Reply:
Should we add `assert(cord_is_main())` to `clock_lowres` then?
@locker

Reply:
No. We just want to ensure that the clock value is written from the same thread. Reading from other threads is fine.

Reply:
But it is data race, shouldn't we use atomic then?

Reply:
As I found out, it is not data race if we have a guarantee that double is atomic.
https://eel.is/c++draft/intro.races#def:data_race

Reply:
Discussed f2f, decided to use timer only from main thread.